### PR TITLE
[Train / Release] Add two checkpoint_upload_fn to async checkpointing benchmark

### PR DESCRIPTION
**Pull request by pseudo-rnd-thoughts**
## Description
`test_async_checkpoint_validation_benchmark.py` is a release test to confirm that ray Train's checkpoint and validation performance to different configurations. 
This PR adds torch distributed checkpoints in sync and async forms based on our [documentation](https://docs.ray.io/en/master/train/user-guides/checkpoints.html#custom-checkpoint-uploading) to confirm implementation performance, both using a custom checkpoint_upload_fn.

This is the reported metrics for each implementation. 
```
{
 'sync_cp_inline_val_metrics': {'e2e_time': 1581.0611889362335,
                                'final_score': 0.5726000070571899,
                                'final_validation_waiting_time': 6.232953310012817,
                                'total_report_blocked_time': 13.458402872085571,
                                'total_validation_time': 203.76652646064758},
 'async_cp_torch_trainer_val_metrics': {'e2e_time': 1389.0721216201782,
                                        'final_score': 0.5835000276565552,
                                        'final_validation_waiting_time': 33.54458975791931,
                                        'total_report_blocked_time': 3.078937292098999,
                                        'total_validation_time': 918.4825644493103},
 'async_cp_map_batches_val_metrics': {'e2e_time': 1362.539043188095,
                                      'final_score': 0.5641,
                                      'final_validation_waiting_time': 23.441231966018677,
                                      'total_report_blocked_time': 3.038931131362915,
                                      'total_validation_time': 426.9236912727356},
  'sync_dcp_map_batches_val_metrics': {'e2e_time': 1357.0950446128845,
                                      'final_score': 0.5753,
                                      'final_validation_waiting_time': 24.127986431121826,
                                      'total_report_blocked_time': 9.00767707824707,
                                      'total_validation_time': 434.1567540168762},
 'async_dcp_map_batches_val_metrics': {'e2e_time': 1361.1273066997528,
                                       'final_score': 0.5758,
                                       'final_validation_waiting_time': 24.674331188201904,
                                       'total_report_blocked_time': 1.38397216796875,
                                       'total_validation_time': 437.4677746295929},
}
```
| Metric                 | Validation Metric Order   | Total E2E Order   | Total Report Block Time Order   | Total Validation Block Time Order   |
|------------------------|---------------------------|-------------------|---------------------------------|-------------------------------------|
| sync_cp_inline         | 2nd (0.57)                | 5th (1581.06)     | 5th (13.46)                     | 5th (223.46)                        |
| async_cp_torch_trainer | 3rd (0.58)                | 4th (1389.07)     | 3rd (3.08)                      | 4th (36.62)                         |
| async_cp_map_batches   | 1st (0.56)                | 3rd (1362.54)     | 2nd (3.04)                      | 2nd (26.48)                         |
| sync_dcp_map_batches   | 4th (0.58)                | 1st (1357.1)      | 4th (9.01)                      | 3rd (33.14)                         |
| async_dcp_map_batches  | 5th (0.58)                | 2nd (1361.13)     | 1st (1.38)                      | 1st (26.06)                         |

### CODE REVIEW

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py [hunk 1]
```diff
@@ -190,15 +227,45 @@ def validate_and_report(
...
+            # Save via Torch DCP
+            dist_cp.save({"model": model_dict}, storage_writer=storage_writer)
+        elif checkpoint_save_mode == CheckpointSaveMode.TORCH_DCP_ASYNC:
+            # Initiate async save; rank 0 will wait via checkpoint_upload_fn
+            ckpt_ref = async_save({"model": model_dict}, storage_writer=storage_writer)
+        else:
+            raise NotImplementedError
+    elif checkpoint_save_mode == CheckpointSaveMode.TORCH_SAVE:
+        # For torch.save, a local temp dir is used and Ray Train handles the upload.
+        iteration_checkpoint_dir = tempfile.mkdtemp()
```
[the remark this answers is not in the document]
Comment (TimothySeah) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: This seems like a real minor issue

File: release/train_tests/async_checkpointing_validation_benchmark/async_checkpointing_validation_benchmark.py [hunk 2]
```diff
@@ -211,12 +278,62 @@ def validate_and_report(
...
+            )
+        elif checkpoint_save_mode == CheckpointSaveMode.TORCH_DCP_ASYNC:
+            # Shards are being written directly to shared storage.
+            # Wait for the async save to finish in a background thread before
+            # marking the checkpoint as ready, but don't upload anything.
+            def wait_async_save(
+                checkpoint, checkpoint_dir_name, upload_complete_ref=ckpt_ref
+            ):
+                upload_complete_ref.result()
+                # todo (mark): this might be unnecessary as the data isn't moving so should be able to use `return checkpoint`
```
Comment (TimothySeah) [inline release/train_tests/async_checkpointing_validation_benchmark/async_checkpointing_validation_benchmark.py]: Can you doublecheck if `return checkpoint` works? That's what the doc example does: https://docs.ray.io/en/latest/train/user-guides/checkpoints.html#asynchronous-checkpoint-uploading

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py [hunk 3]
```diff
@@ -211,12 +278,62 @@ def validate_and_report(
...
+                checkpoint, checkpoint_dir_name, upload_complete_ref=ckpt_ref
+            ):
+                upload_complete_ref.result()
+                # todo (mark): this might be unnecessary as the data isn't moving so should be able to use `return checkpoint`
+                path = (
+                    ray.train.get_context()
+                    .get_storage()
+                    .build_checkpoint_path_from_name(checkpoint_dir_name)
+                )
+                return ray.train.Checkpoint.from_directory(path)
```
[the remark this answers is not in the document]
Comment (TimothySeah) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: Wait is this a real issue? I thought every worker calls `checkpoint_upload_fn` and waits.

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py [hunk 4]
```diff
@@ -5,13 +5,15 @@
 import time
 
 import torch
+import torch.distributed.checkpoint as dist_cp
 import torchmetrics
+from torch.distributed.checkpoint.state_dict import get_state_dict
```
Comment (TimothySeah) [inline release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py:10]: What's the main blocker for testing with direct s3 uploads?

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py [hunk 5]
```diff
@@ -211,12 +278,62 @@ def validate_and_report(
...
+
+            ray.train.report(
+                metrics,
+                checkpoint=ray.train.Checkpoint.from_directory(
+                    iteration_checkpoint_dir
+                ),
+                checkpoint_upload_fn=wait_async_save,
+                checkpoint_dir_name=f"dcp_epoch_{epoch}_batch_{batch_idx}",
+                checkpoint_upload_mode=CheckpointUploadMode.ASYNC,
+                # iteration_checkpoint_dir is already in shared storage so don't delete it.
```
Comment (TimothySeah) [inline release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py:327]: Good catch. Can you also update https://docs.ray.io/en/latest/train/user-guides/checkpoints.html#asynchronous-checkpoint-uploading to do `delete_local_checkpoint_after_upload=False`?

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py [hunk 6]
```diff
@@ -361,9 +494,34 @@ def main():
...
+        False,
+        num_epochs,
+        train_dataset,
+        training_rows,
+        CheckpointSaveMode.TORCH_DCP_ASYNC,
     )
     logger.info(consolidated_metrics)
     safe_write_to_results_json(consolidated_metrics)
+    for run_name, metrics in consolidated_metrics.items():
+        print(f"{run_name}={metrics}")
```
Comment (TimothySeah) [inline release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: nit: logger instead of print. And maybe you can remove the `logger.info` call above?

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py [hunk 7]
```diff
@@ -374,12 +532,32 @@ def main():
     async_map_batches_final_score = consolidated_metrics[
         "async_cp_map_batches_val_metrics"
     ]["final_score"]
+    sync_dcp_final_score = consolidated_metrics["sync_dcp_map_batches_val_metrics"][
+        "final_score"
+    ]
+    async_dcp_final_score = consolidated_metrics["async_dcp_map_batches_val_metrics"][
+        "final_score"
+    ]
+    print(
```
Comment (TimothySeah) [inline release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: nit: logger instead of print here too

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py [hunk 8]
```diff
@@ -390,21 +568,42 @@ def main():
     async_map_batches_e2e_time = consolidated_metrics[
         "async_cp_map_batches_val_metrics"
     ]["e2e_time"]
+    sync_dcp_e2e_time = consolidated_metrics["sync_dcp_map_batches_val_metrics"][
+        "e2e_time"
+    ]
+    async_dcp_e2e_time = consolidated_metrics["async_dcp_map_batches_val_metrics"][
+        "e2e_time"
+    ]
+    print(
```
Comment (TimothySeah) [inline release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: logger

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py (hunk 7 above)
Comment (TimothySeah) [inline release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: You added a bunch of print statements to the code - can you also copy and paste what they look like to the PR description?

**Review by TimothySeah — COMMENTED** [review]
Mostly LGTM - I probably just need 1 more quick pass after this one.

File: release/train_tests/async_checkpointing_validation_benchmark/async_checkpointing_validation_benchmark.py (hunk 2 above)
Comment (pseudo-rnd-thoughts) [reply release/train_tests/async_checkpointing_validation_benchmark/async_checkpointing_validation_benchmark.py]: Yes, I will test it

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py (hunk 4 above)
Comment (pseudo-rnd-thoughts) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py:10]: Our experiment directory is local while the checkpoint would be on S3 and we don't support this type of out of band checkpointing currently. We could update the experiment directory to use S3 but I haven't done that yet.

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py (hunk 5 above)
Comment (pseudo-rnd-thoughts) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py:327]: Yes, will do

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py (hunk 6 above)
Comment (pseudo-rnd-thoughts) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: Ahh, I missed the log above, yes, I will update to log

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py (hunk 7 above)
Comment (pseudo-rnd-thoughts) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: will do

Comment (pseudo-rnd-thoughts) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: Will do

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: Sadly, yes. `ckpt_ref.result()` will ensure that this worker is complete not that all workers are and that PyTorch recommends using a barrier after async_save to ensure that all workers have completed it. 
I'm not sure how much we want to change our implementation as it would require use to call `ray.train.report` on all workers.
What are your thoughts?

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py (hunk 5 above)
Comment (pseudo-rnd-thoughts) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py:327]: Added to https://github.com/ray-project/ray/pull/62555

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TimothySeah) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: Iiuc all workers already call `ray.train.report` which has a barrier so this should be fine, but please doublecheck my reasoning.

In this release test:
* Every worker calls `async_save`: https://github.com/pseudo-rnd-thoughts/ray/blob/9af477046a08314a3624198704f6dce3b8b27562/release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py#L256, which kicks off an async checkpoint upload. Note: actually we can probably limit this to rank 0 anyway.
* Every worker calls `ray.train.report` (https://github.com/pseudo-rnd-thoughts/ray/blob/9af477046a08314a3624198704f6dce3b8b27562/release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py#L322) with a custom checkpoint upload function that waits for its async checkpoint upload to complete. Note: once again, only rank 0 needs to do this anyway, but the other workers still need to make an empty report (look at the examples here: https://docs.ray.io/en/latest/train/user-guides/checkpoints.html#saving-checkpoints-during-training). 
* The async report call kicks off a new thread that uploads the checkpoint and then "reports" it by adding it to a queue on the worker (https://github.com/ray-project/ray/blob/f9ccc7a79ee4535a5575551687e193c003e6c7f9/python/ray/train/v2/_internal/execution/context.py#L445). Note that this queue serves as a barrier in the sense that all workers must report something (https://github.com/ray-project/ray/blob/f9ccc7a79ee4535a5575551687e193c003e6c7f9/python/ray/train/v2/_internal/execution/checkpoint/report_handler.py#L63) before the train controller kicks off validation for it through the `after_report` callback: https://github.com/ray-project/ray/blob/f9ccc7a79ee4535a5575551687e193c003e6c7f9/python/ray/train/v2/_internal/execution/checkpoint/report_handler.py#L98

[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: I got claude to review the actual `async_save` to confirm cursor's and my understanding of the documentation and thankfully my prior understanding was wrong.  

> async_save runs dist_cp.save() in a background thread on all workers. Internally, save() ends with an all_reduce across all ranks:
>  1. Each rank calls write_data() (writes local shards, waits for them to finish)
>  2. The all_reduce collects write results from ALL ranks to the coordinator
>  3. Only then does rank 0 call finish_checkpoint() and write the .metadata file                                                                                   
> So rank 0's future cannot resolve until every rank has finished writing its shards, it's guaranteed by the all_reduce. When wait_async_save calls upload_complete_ref.result() on rank 0's future, the checkpoint is fully written.

Thankfully this means our current code is completely fine

**Review by TimothySeah — APPROVED** [review]
Can you doublecheck the math? I would expect async dcp map batches to be the fastest. Other than that LGTM.
